### ISSUE TITLE
Discussion: Extending NIC selectors

### ISSUE THREAD

**Issue by martinkennelly** [k8snetworkplumbingwg/sriov-network-operator#60]
I propose we extend the ```sriovNetworkNicSelector``` fields to include additional optional selectors as defined in SRIOV DP. I want to generate discussion which additional selectors we would like to include.

For my customer use case, I want the SRIOV Network Operator to be able to group VFs by DDP Profile. This would involve exposing the ```ddpProfiles``` as an optional selector in this Operator. The provisioning of DDP on a NIC will be a pre-requisite outside the scope of this Operator. 

I raised this topic at the additional network & resource community meeting and got positive feedback of adding new optional NIC selectors.

**Comment by pliurh** [comment]
@martinkennelly I'm open to add this ddp selector to the SRIOV operator. However, we also need to think about how to discover the ddp profiles on different NIC models.

**Comment by ipatrykx** [comment]
Hi @zshi-redhat @pliurh I would like to continue the topic started by @martinkennelly.  I have prepared small overview of what we would like to achieve and what changes would be required.

What could be added:

Support to use DDP profile name as a selector, what is currently supported by SRIOV Network Device Plugin itself: https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/tree/master/docs/ddp

Justification:

This feature is currently supported by DP. However it seems that SRIOV Network Operator does not support it as the team does not want to incorporate ‘DDPTool’ binary into Network Operator. However, we will soon introduce the changes to the DP community that will make it possible to read the DDP Package name using devlink directly for Intel’s E810 (Columbiaville) NICs (and other NICs that support’s devlink). Hopefully those changes will be accepted by the community.

Proposed workflow:

1. Add support for devlink in SRIOV Network Device Plugin (implementation done, minor tweaks needed). Two possible ways of doing this (done simultaneously):
	- Add Devlink related logic to SRIOV DP
	- Add Devlink related logic to vishvananda/netlink library (preferred)
2. Add proper NIC selector to SRIOV Network Operator
	- Add ddpProfiles to SriovNetworkNicSelector
  	- util.go -> validateSelector - validate if vendor == Intel && card == CVL
3. This should be automatically parsed to sriov-device-plugin configmap. Add tests.

Out of scope:

Provisioning of the DDP package to the NIC (flashing) is out of scope for this changes. It would be nice to implement that, so node daemon would do that simply using SriovNetworkNodeState but that would require a way to find package file on the node. As I believe it would be possible to achieve that, I would like to discuss this after the selector itself will be supported.

**Comment by adrianchiris** [comment]
@ipatrykx can you share some more info on the devlink work ?
are those devlink device parameters ?

**Comment by ipatrykx** [comment]
@adrianchiris I think I will be able to create PR on DP in few days. Anyway, it uses devlink-info API, which general overview is provided [in Kernel's docs](https://www.kernel.org/doc/html/v5.8/networking/devlink/devlink-info.html) and CVL specific implementation can be found [here](https://www.kernel.org/doc/html/v5.8/networking/devlink/ice.html).

From DP point of view it's just about issuing devlink's `DEVLINK_CMD_INFO_GET` command and parsing the response.

**Comment by ipatrykx** [comment]
Hi all, just to refresh this thread a bit -I already have this feature implemented, however I am waiting for the internal QA process to be finished.

**Comment by ognjen011** [comment]
Just to find out has this ever been done can we use the ddpProfiles?

**Comment by SchSeba** [comment]
Hi @ognjen011 this was not implemented in the operator because we didn't get any requirement for it.

be aware that we will need to first revive https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pull/387 before we can add support in the operator
